From patchwork Thu Jan  8 18:03:15 2026
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Tianlan Zhou <bobby825@126.com>
X-Patchwork-Id: 2181561
Return-Path: <qemu-devel-bounces+incoming=patchwork.ozlabs.org@nongnu.org>
X-Original-To: incoming@patchwork.ozlabs.org
Delivered-To: patchwork-incoming@legolas.ozlabs.org
Authentication-Results: legolas.ozlabs.org;
	dkim=pass (1024-bit key;
 unprotected) header.d=126.com header.i=@126.com header.a=rsa-sha256
 header.s=s110527 header.b=TfZVNAXn;
	dkim-atps=neutral
Authentication-Results: legolas.ozlabs.org;
 spf=pass (sender SPF authorized) smtp.mailfrom=nongnu.org
 (client-ip=209.51.188.17; helo=lists.gnu.org;
 envelope-from=qemu-devel-bounces+incoming=patchwork.ozlabs.org@nongnu.org;
 receiver=patchwork.ozlabs.org)
Received: from lists.gnu.org (lists.gnu.org [209.51.188.17])
	(using TLSv1.2 with cipher ECDHE-ECDSA-AES256-GCM-SHA384 (256/256 bits))
	(No client certificate requested)
	by legolas.ozlabs.org (Postfix) with ESMTPS id 4dnCTz4H39z1xqr
	for <incoming@patchwork.ozlabs.org>; Fri, 09 Jan 2026 05:04:39 +1100 (AEDT)
Received: from localhost ([::1] helo=lists1p.gnu.org)
	by lists.gnu.org with esmtp (Exim 4.90_1)
	(envelope-from <qemu-devel-bounces@nongnu.org>)
	id 1vduMc-0005W4-MQ; Thu, 08 Jan 2026 13:04:03 -0500
Received: from eggs.gnu.org ([2001:470:142:3::10])
 by lists.gnu.org with esmtps (TLS1.2:ECDHE_RSA_AES_256_GCM_SHA384:256)
 (Exim 4.90_1) (envelope-from <bobby825@126.com>) id 1vduMS-00053G-DC
 for qemu-devel@nongnu.org; Thu, 08 Jan 2026 13:03:55 -0500
Received: from m16.mail.126.com ([117.135.210.9])
 by eggs.gnu.org with esmtps (TLS1.2:ECDHE_RSA_AES_256_GCM_SHA384:256)
 (Exim 4.90_1) (envelope-from <bobby825@126.com>) id 1vduMN-0003ZV-De
 for qemu-devel@nongnu.org; Thu, 08 Jan 2026 13:03:52 -0500
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=126.com;
 s=s110527; h=From:To:Subject:Date:Message-ID:MIME-Version; bh=09
 tCkdU8nwORaDi7FEX5Ds2RD2ozXaRRISY7ZYG2FPM=; b=TfZVNAXnmJs9k5/QHy
 JQqRaWbr36ahszJkTxdHkL04PSfCjhvdmrm4E89E3wFqvon4LMvsn4gdhH0L0j4x
 infaUtQHpaoNSIXTroaWl3uGhXfY63z9hUlG0/aRkIxZNwvtvwvscQSOU8Ac2TS0
 adwJSSjAPDLXHyQWMXsb1xtnY=
Received: from SSS-WIN-GP76HX (unknown [])
 by gzga-smtp-mtada-g1-4 (Coremail) with SMTP id
 _____wDXv7dn8V9pt_Q6Bg--.22025S2;
 Fri, 09 Jan 2026 02:03:20 +0800 (CST)
From: Tianlan Zhou <bobby825@126.com>
To: qemu-devel@nongnu.org
Cc: Tianlan Zhou <bobby825@126.com>,
	Stefan Weil <sw@weilnetz.de>
Subject: [PATCH] win32: Add ftruncate function detection
Date: Fri,  9 Jan 2026 02:03:15 +0800
Message-ID: <20260108180315.1587-1-bobby825@126.com>
X-Mailer: git-send-email 2.52.0.windows.1
MIME-Version: 1.0
X-CM-TRANSID: _____wDXv7dn8V9pt_Q6Bg--.22025S2
X-Coremail-Antispam: 1Uf129KBjvJXoW7KF47tw17Xr48Wr15urW5trb_yoW8GrWfpF
 Z2yryYgF4UJ34rJrWxuF1xur4fXF4kGw17GFyUtr1kWr1xXF1Utrs8K3Wakw47XrsYyr4r
 uFyfJFZxurWjqr7anT9S1TB71UUUUU7qnTZGkaVYY2UrUUUUjbIjqfuFe4nvWSU5nxnvy2
 9KBjDUYxBIdaVFxhVjvjDU0xZFpf9x0pE73kXUUUUU=
X-Originating-IP: [2409:8a1e:6e71:2040:a1d7:92ec:eb8e:be55]
X-CM-SenderInfo: xereu5qysvqiyswou0bp/xtbBqAnIG2lf8WlxEwAA3+
Received-SPF: pass client-ip=117.135.210.9; envelope-from=bobby825@126.com;
 helo=m16.mail.126.com
X-Spam_score_int: -17
X-Spam_score: -1.8
X-Spam_bar: -
X-Spam_report: (-1.8 / 5.0 requ) BAYES_00=-1.9, DKIM_SIGNED=0.1,
 DKIM_VALID=-0.1, DKIM_VALID_AU=-0.1, DKIM_VALID_EF=-0.1,
 FREEMAIL_ENVFROM_END_DIGIT=0.25, FREEMAIL_FROM=0.001,
 RCVD_IN_DNSWL_NONE=-0.0001, RCVD_IN_MSPIKE_H5=0.001, RCVD_IN_MSPIKE_WL=0.001,
 RCVD_IN_VALIDITY_RPBL_BLOCKED=0.001, RCVD_IN_VALIDITY_SAFE_BLOCKED=0.001,
 SPF_HELO_NONE=0.001, SPF_PASS=-0.001,
 UNPARSEABLE_RELAY=0.001 autolearn=ham autolearn_force=no
X-Spam_action: no action
X-BeenThere: qemu-devel@nongnu.org
X-Mailman-Version: 2.1.29
Precedence: list
List-Id: <qemu-devel.nongnu.org>
List-Unsubscribe: <https://lists.nongnu.org/mailman/options/qemu-devel>,
 <mailto:qemu-devel-request@nongnu.org?subject=unsubscribe>
List-Archive: <https://lists.nongnu.org/archive/html/qemu-devel>
List-Post: <mailto:qemu-devel@nongnu.org>
List-Help: <mailto:qemu-devel-request@nongnu.org?subject=help>
List-Subscribe: <https://lists.nongnu.org/mailman/listinfo/qemu-devel>,
 <mailto:qemu-devel-request@nongnu.org?subject=subscribe>
Errors-To: qemu-devel-bounces+incoming=patchwork.ozlabs.org@nongnu.org
Sender: qemu-devel-bounces+incoming=patchwork.ozlabs.org@nongnu.org

Add detection for `ftruncate` function in Meson, so that it can be
used when possible instead of fallback version `qemu_ftruncate64`.

Signed-off-by: Tianlan Zhou <bobby825@126.com>
Resolves: https://gitlab.com/qemu-project/qemu/-/issues/3234
---
 include/system/os-win32.h |  2 +-
 meson.build               | 11 +++++++++++
 2 files changed, 12 insertions(+), 1 deletion(-)

diff --git a/include/system/os-win32.h b/include/system/os-win32.h
index 22d72babdf..9832b7589a 100644
--- a/include/system/os-win32.h
+++ b/include/system/os-win32.h
@@ -140,7 +140,7 @@ static inline void os_setup_limits(void)
 
 int qemu_ftruncate64(int, int64_t);
 
-#if !defined(ftruncate)
+#ifndef CONFIG_FTRUNCATE
 # define ftruncate qemu_ftruncate64
 #endif
 
diff --git a/meson.build b/meson.build
index db87358d62..0b41952a0b 100644
--- a/meson.build
+++ b/meson.build
@@ -3253,6 +3253,17 @@ if host_os == 'windows'
   config_host_data.set('CONFIG_CONVERT_STRING_TO_BSTR', has_convert_string_to_bstr)
 endif
 
+# Detect if ftruncate exist
+if host_os == 'windows'
+  has_ftruncate = cxx.links('''
+    #include <unistd.h>
+    int main() {
+        return ftruncate(0, 0);
+    }
+  ''')
+  config_host_data.set('CONFIG_FTRUNCATE', has_ftruncate)
+endif
+
 ########################
 # Target configuration #
 ########################
